csrc/mmdeploy/core/registry.cpp

```diff
@@ -2,53 +2,80 @@
 
 #include "mmdeploy/core/registry.h"
 
-namespace mmdeploy {
+#include <algorithm>
+#include <cassert>
 
-Registry<void>::Registry() = default;
+#include "mmdeploy/core/logger.h"
 
-Registry<void>::~Registry() = default;
+namespace mmdeploy {
 
-bool Registry<void>::AddCreator(Creator<void>& creator) {
-  MMDEPLOY_DEBUG("Adding creator: {}", creator.GetName());
-  auto key = creator.GetName();
-  if (entries_.find(key) == entries_.end()) {
-    entries_.insert(std::make_pair(key, &creator));
-    return true;
-  }
+namespace _registry {
 
-  for (auto iter = entries_.lower_bound(key); iter != entries_.upper_bound(key); ++iter) {
-    if (iter->second->GetVersion() == creator.GetVersion()) {
-      return false;
-    }
+struct Registry<void>::Impl {
+  template <typename It>
+  auto convert(It u, It v) {
+    return std::pair{creators_.begin() + (u - names_.begin()),
+                     creators_.begin() + (v - names_.begin())};
   }
 
-  entries_.insert(std::make_pair(key, &creator));
-  return true;
-}
-
-Creator<void>* Registry<void>::GetCreator(const std::string& type, int version) {
-  auto iter = entries_.find(type);
-  if (iter == entries_.end()) {
+  Creator<void>* Get(const string_view& name, int version) {
+    const auto& [u, v] = std::equal_range(names_.begin(), names_.end(), name);
+    const auto& [i, j] = convert(u, v);
+    if (version == -1) {
+      if (auto n = j - i; n == 1) {
+        return *i;
+      }
+      return nullptr;
+    }
+    for (const auto& creator : iterator_range(i, j)) {
+      if (creator->version() == version) {
+        return creator;
+      }
+    }
     return nullptr;
   }
-  if (0 == version) {
-    return iter->second;
-  }
 
-  for (auto iter = entries_.lower_bound(type); iter != entries_.upper_bound(type); ++iter) {
-    if (iter->second->GetVersion() == version) {
-      return iter->second;
+  bool Add(Creator<void>& creator) {
+    const auto& [u, v] = std::equal_range(names_.begin(), names_.end(), creator.name());
+    const auto& [i, j] = convert(u, v);
+    if (i != j) {
+      for (const auto& other : iterator_range(i, j)) {
+        if (creator.version() == other->version()) {
+          MMDEPLOY_WARN("Adding duplicated creator ({}, {}).", creator.name(), creator.version());
+          return false;
+        }
+      }
     }
+    names_.insert(v, creator.name());
+    creators_.insert(j, &creator);
+    return true;
   }
-  return nullptr;
+
+  Span<Creator<void>*> Creators() { return creators_; }
+
+  std::vector<Creator<void>*> creators_;
```

Comment:
Should we put the member variables in private part?

Reply:
Why use a vector to cache the creators? Is it better than `std::map`?
Everytime a new creator is added, the vector is probably re-arranged, isn't it?

Reply:
> Should we put the member variables in private part?

It does not matter, `struct Registry<void>::Impl` is private and hidden anyway.

> Why use a vector to cache the creators? Is it better than `std::map`?

It's simply a implementation of a `flat_multimap`, benchmark on `std::multimap` and `flat_multimap` reveals that if you have less than 1000 items, insertion on the flat variant is faster. And `std::equal_range` is always faster than `std::multimap::equal_range` due to better locality.